src/TextInputWidget.cc

```diff
@@ -101,6 +111,51 @@ FilteredTextEdit::keyPressEvent(QKeyEvent *event)
         }
 }
 
+bool
+FilteredTextEdit::canInsertFromMimeData(const QMimeData *source) const
+{
+        return (source->hasImage() || source->hasText() ||
+                QTextEdit::canInsertFromMimeData(source));
+}
+
+void
+FilteredTextEdit::insertFromMimeData(const QMimeData *source)
+{
+        if (source->hasImage()) {
+                const QImage image = qvariant_cast<QImage>(source->imageData());
```

Comment:
On further investigation, it's apparent to me that we can and therefore must use [`QMimeData::data`](https://doc.qt.io/qt-5/qmimedata.html#data) to avoid re-encoding the image at all, rather than decoding the image here and re-encoding it below. This will avoid silent quality loss, improve robustness, and reduce CPU and memory use. The mime type obtained from `QMimeData` should also be propagated alongside the data so it can be included in the matrix event verbatim, removing any need to guess it.

I apologize for not identifying this option earlier; I'm not super familiar with how rich copy/paste works.

Reply:
Forgive me if I'm missing something here, but using `QMimeData::data` won't prevent re-encoding, because we must pass a `QString` representing the MIME type, which is then used to create (encode the raw pixels) the `QByteArray`. I also am not too familiar with copy/paste media, but I've noticed that when copying an image (not from a file manager), it will copy the raw pixel bytes, understandably, as there's no notion of a file. I only came to this realization when creating this PR. 

Therefore, re-encoding cannot be avoided. However, your suggestion has reduced the amount of code necessary and made it much cleaner.

Reply:
> we must pass a `QString` representing the MIME type, which is then used to create (encode the raw pixels) the `QByteArray`

My understanding is that `QMimeData::data` performs no transcoding. The supplied `mimeType` is used to select *which* of multiple pieces of attached data you want (from the alternatives exposed by `formats`). This can be observed in the [`QMimeDataPrivate::retrieveTypedData`](https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qmimedata.cpp#n108) implementation's  handling of the `type == QVariant::ByteArray` case, which for image data falls through ultimately to [`QMimeDataPrivate::getData`](https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qmimedata.cpp#n96), which is a simple linear search through `dataList`.

> I also am not too familiar with copy/pasting media, but I've noticed that when copying an image (not from a file manager), it will copy the raw pixel bytes, understandably, as there's no notion of a file.

What was your methodology here? If we're genuinely receiving `image/*` mime types, as the current PR suggests, I would expect the data to be in the advertised format. In fact, reviewing the implementation, it appears to me that [`QMimeData::imageData` returns nonempty results *only* when the mime type is `application/x-qt-image`](https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qmimedata.cpp#n478). PNG data certainly doesn't have to be physically stored in a file to be valid.

*time passes*

Right clicking on a jpg image in Chrome, selecting "Copy image," and inspecting the clipboard state with `xclip -o -selection clipboard -t TARGETS` reveals that the clipboard contains formats, `image/png` and `text/html`. Dumping the former yields a valid png image, and the latter yields a brief HTML document containing an image tag. I'm a bit perplexed by Chrome's apparent decision to transcode the image, but it's apparent that we do not need to.

> re-encoding cannot be avoided

If uncompressed data is actually the norm, we should take care to encode it as png (or perhaps something more modern, like webp?) for sanity's sake, but per the above testing this does not appear to be the case.

Reply:
Huh, seems I have drawn an incorrect conclusion because of a misinterpretation of the `QImage::bits` method. These are the steps that I took which led me to make my initial claim:

```cpp
QImage img = qvariant_cast<QImage>(source->imageData());
QByteArray arr = QByteArray{img.bits(), img.byteCount()};
```
For some reason, I thought that this would essentially convert the `QImage` into a byte array. Instead, it gives the actual raw pixel bytes, which led me to incorrectly believe that "Copy image" copies raw bytes even after saving the `QImage` as PNG.

Alright well, I apologize for that, my knowledge of the clipboard is clearly lacking. I was not even aware of that particular capability of `xclip`, that's pretty nifty.

Reply:
I think `QImage` probably *is* uncompressed data, since it supports editing. I'm still a bit confused about whether `imageData` is ever nonempty when pasting e.g. `image/png`; it seems like it can't be, but I assume you've the original code and it does something at least some of the time. Perhaps Qt automatically decodes incoming pastes for convenience of display.

One potential subtlety to bear in mind is the case where many image formats available. For example, here's the list firefox supplies:
```
image/png
image/bmp
image/x-bmp
image/x-MS-bmp
image/jpeg
image/tiff
image/x-icon
image/x-ico
image/x-win-bitmap
image/vnd.microsoft.icon
application/ico
image/ico
image/icon
text/ico
```
Given that the data supplied by both firefox and chrome seems to be re-encoded (the input was a jpeg image, and the image/jpeg content was 6 times smaller than the input) IMO a lossless format should be preferred. I'm happy to work under the hypothesis that everyone's helpful and orders paste data in descending quality--i.e. that it is correct to pick the first `image` content if any exists--but if this is determined to often be false, we may in the future want to add some heuristics to try to make a good selection. Might be worth leaving a note to ourselves in a comment.